### PR TITLE
Perform silent expr checks in QueryTypeChecker when resolving selectTypes

### LINKED ISSUE #40012: [Bug]: Compilation fails when using query expressions to assign values for union typed record fields
### Description

Getting the following compilation error when I'am trying to assign values to union typed record field using query expression.
```
ambiguous type '[org/project:0.1.0:T3, org/project:0.1.0:T3]'
```

### Steps to Reproduce

Sample source:
```ballerina

type T1 record {
    T3[] t3s;
};

type T2 record {
    T3[]|T4[] t3OrT4;
};

type T3 record {
    string str;
};

type T4 record {
    boolean foo;
};

function transform(T1 t1) returns T2 => {
    t3OrT4: from var t3sItem in t1.t3s
        select {
           str: ""
        }
};
```

The error remains even I did a type cast within the select clause
```ballerina
function transform(T1 t1) returns T2 => {
    t3OrT4: from var t3sItem in t1.t3s
        select <T3>{
           str: ""
        }
};
```

### Affected Version(s)

Ballerina `2201.4.0`

### OS, DB, other environment details and versions

_No response_

### Related area

-> Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by sanjana**
Fixes #40012

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [x] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Comment by LakshanWeerasinghe** [comment]
There are checkstyle errors in the `TypeChecker`.

**Comment by sanjana** [comment]
> There are checkstyle errors in the `TypeChecker`.

Addressed in https://github.com/ballerina-platform/ballerina-lang/pull/40271/commits/1d8ef667fab06538e131584875b982ce81700782.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/QueryTypeChecker.java [hunk 1]
```diff
@@ -288,79 +289,116 @@ public BType resolveQueryType(SymbolEnv env, BLangExpression selectExp, BType ta
...
+                        continue;
+                    }
+                    resolvedType = getResolvedType(symTable.tableType, type, isReadonly, env);
+                    break;
+                case TypeTags.STREAM:
+                    BType streamConstraint = types.getSafeType(((BStreamType) type).constraint,
+                            true, true);
+                    selectType = checkExprSilent(selectExp, env, streamConstraint, data);
+                    if (selectType.tag == TypeTags.SEMANTIC_ERROR) {
+                        errorTypes.add(selectType);
```
Comment (pcnfernando) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/QueryTypeChecker.java:327]: Here, `streamConstraint` should be added if the `checkExpr` returns an erroneous type

Comment (pcnfernando) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/QueryTypeChecker.java:327]: Add a test to cover each of this erronous type

**Comment by pcnfernando** [comment]
The build is failing with a check style issue

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/QueryTypeChecker.java (hunk 1 above)
Comment (sanjana) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/QueryTypeChecker.java]: Added the `streamConstraint` to errorTypes to perform the `checkExpr` with test cases in https://github.com/ballerina-platform/ballerina-lang/pull/40271/commits/247e2b02e54ee8b97bef088ef2f9a38d18d88ba7.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/QueryTypeChecker.java [hunk 2]
```diff
@@ -249,14 +249,15 @@ public BType resolveQueryType(SymbolEnv env, BLangExpression selectExp, BType ta
         List<BType> resolvedTypes = new ArrayList<>();
         BType selectType;
         BLangExpression collectionNode = (BLangExpression) ((BLangFromClause) clauses.get(0)).getCollection();
-        for (BType type : safeResultTypes) {
-            solveSelectTypeAndResolveType(queryExpr, selectExp, type, collectionNode.getBType(), selectTypes,
-                    resolvedTypes, env, data, false);
-        }
+        solveSelectTypeAndResolveType(queryExpr, selectExp, safeResultTypes, collectionNode.getBType(), selectTypes,
+                resolvedTypes, env, data, false);
+
```
[the remark this answers is not in the document]
Comment (sanjana) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/QueryTypeChecker.java:254]: The information such as the `expTypes` count should be accessible within the method to handle union types as expected types.

**Comment by sanjana** [comment]
> The build is failing with a check style issue. @sanjana need your immediate attention to get this merged asap

@pcnfernando updated the PR by addressing the updates and testing the unit test cases locally.

**Comment by pcnfernando** [comment]
@sanjana builds are failing

**Comment by sanjana** [comment]
> @sanjana builds are failing

Re-synced with the master and re-triggered the build.

**Comment by pcnfernando** [comment]
@sanjana please check on the checkstyle failure

**Review by LakshanWeerasinghe — DISMISSED** [review]
Changed looks good except for the case where there can be in a union type inside the select clause. In that case when we are resolving the select clause type and when we have more than one type is resolved as error we have to give the correct actual and expected type and type check the full query expression.

This will be fixed when we fix #40013.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 3]
```diff
@@ -3645,6 +3645,14 @@ protected BType checkExprSilent(BLangExpression expr, BType expType, AnalyzerDat
         return type;
     }
 
+    protected BType checkExprSilent(BLangExpression expr, SymbolEnv env, BType expType, AnalyzerData data) {
+        SymbolEnv prevEnv = data.env;
+        data.env = env;
+        BType type = checkExprSilent(nodeCloner.cloneNode(expr), expType, data);
```
[the remark this answers is not in the document]
Comment (SasinduDilshara) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:3651]: Shall we do the cloning operation when we calling this function, like the same method we use

[the remark this answers is not in the document]
Comment (sanjana) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:3651]: This approach was used to minimize the redundant calls to the `nodeCloner.cloneNode()` whenever the silent checks are being performed with the overloaded SymbolEnv param.

**Review by SasinduDilshara — APPROVED** [review]

**Pull request merged**